### PR TITLE
krun: init at 0-unstable-2024-06-18

### PR DESCRIPTION
**Pull request by RossComputerGuy**
## Description of changes

Updates `libkrun` and `libkrunfw`, adds `krun`. Good way to test is `NIXOS_CURRENT_SYSTEM=$(readlink -e /run/current-system/) ./result/bin/krun bash`

## Things done

- Built on platform(s)
  - [x] x86_64-linux
  - [x] aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
- For non-Linux: Is sandboxing enabled in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
  - [ ] `sandbox = relaxed`
  - [ ] `sandbox = true`
- [ ] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://github.com/NixOS/nixpkgs/blob/master/pkgs/README.md#package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://nixos.org/manual/nixpkgs/unstable/#ssec-nixos-tests-linking) to the relevant packages
- [ ] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [x] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [24.11 Release Notes](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2411.section.md) (or backporting [23.11](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2311.section.md) and [24.05](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2405.section.md) Release notes)
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

---

Add a :+1: [reaction] to [pull requests you find important].

[reaction]: https://github.blog/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/
[pull requests you find important]: https://github.com/NixOS/nixpkgs/pulls?q=is%3Aopen+sort%3Areactions-%2B1-desc

### CODE REVIEW

File: pkgs/by-name/kr/krun/package.nix [hunk 1]
```diff
@@ -0,0 +1,82 @@
...
+, opengl-driver ? mesa.drivers
+, withSommelier ? false
+}:
+
+stdenv.mkDerivation (finalAttrs: {
+  pname = "krun";
+  version = "0.0.0-unstable-2024-06-18";
+
+  src = fetchFromGitHub {
+    owner = "slp";
```
Comment (flokli) [inline pkgs/by-name/kr/krun/package.nix:23]: https://github.com/slp/krunvm seems to be outdated (last commit on `main` 2 years ago), while forked off https://github.com/containers/krunvm, which was last touched two weeks ago (by `slp`). Are you sure the upstream is correct?

File: pkgs/by-name/kr/krun/package.nix [hunk 2]
```diff
@@ -0,0 +1,82 @@
...
+  ];
+
+  postFixup = ''
+    wrapProgram $out/bin/krun $wrapArgs \
+      --set-default OPENGL_DRIVER ${opengl-driver}
+  '';
+
+  meta = {
+    description = "krun - run programs from your system in a microVM";
+    homepage = "https://github.com/slp/krunvm";
```
Comment (flokli) [inline pkgs/by-name/kr/krun/package.nix:76]: same

**Review by flokli — DISMISSED** [review]

File: pkgs/by-name/kr/krun/package.nix (hunk 1 above)
Comment (RossComputerGuy) [reply pkgs/by-name/kr/krun/package.nix:23]: krun != krunvm

File: pkgs/by-name/kr/krun/package.nix (hunk 2 above)
Comment (RossComputerGuy) [reply pkgs/by-name/kr/krun/package.nix:76]: Forgot to update the homepage URL correctly

Comment (RossComputerGuy) [reply pkgs/by-name/kr/krun/package.nix]: Fixed

File: pkgs/by-name/kr/krun/package.nix (hunk 1 above)
Comment (RossComputerGuy) [reply pkgs/by-name/kr/krun/package.nix:23]: Ref: https://github.com/containers/krunvm/issues/62#issuecomment-2135123678

Comment (RossComputerGuy) [reply pkgs/by-name/kr/krun/package.nix:23]: The differences between krun and krunvm are: krunvm uses OCI images and relies on buildah to work, it runs a VM. krun is a VM sandbox of your host OS and wraps things in a nice environment for systems like Apple Silicon.

File: pkgs/by-name/kr/krun/package.nix [hunk 3]
```diff
@@ -0,0 +1,82 @@
...
+, passt
+, sommelier
+, mesa
+, opengl-driver ? mesa.drivers
+, withSommelier ? false
+}:
+
+stdenv.mkDerivation (finalAttrs: {
+  pname = "krun";
+  version = "0.0.0-unstable-2024-06-18";
```
Comment (Aleksanaa) [inline pkgs/by-name/kr/krun/package.nix:20]:
```suggestion
  version = "0-unstable-2024-06-18";
```

Why's that `0.0.0`?

File: pkgs/by-name/kr/krun/package.nix [hunk 4]
```diff
@@ -0,0 +1,82 @@
...
+    "--prefix PATH : ${lib.makeBinPath (lib.optional withSommelier [ sommelier ] ++ [ passt ])}"
+  ];
+
+  postFixup = ''
+    wrapProgram $out/bin/krun $wrapArgs \
+      --set-default OPENGL_DRIVER ${opengl-driver}
+  '';
+
+  meta = {
+    description = "krun - run programs from your system in a microVM";
```
Comment (Aleksanaa) [inline pkgs/by-name/kr/krun/package.nix:75]:
```suggestion
    description = "Run programs from your system in a microVM";
```

File: pkgs/by-name/kr/krun/package.nix [hunk 5]
```diff
@@ -0,0 +1,82 @@
...
+    })
+  ];
+
+  cargoDeps = rustPlatform.fetchCargoTarball {
+    inherit (finalAttrs) src;
+    hash = "sha256-efJvEpmSBTrShOc9fwmn2TqC1zYx5WO4LH/6JQtIErA=";
+  };
+
+  nativeBuildInputs = [
+    rustPlatform.cargoSetupHook
```
Comment (Aleksanaa) [inline pkgs/by-name/kr/krun/package.nix:46]: So we normally just use `rustPlatform.buildRustPackage` (Is there anything wrong with that?)

File: pkgs/by-name/kr/krun/package.nix (hunk 3 above)
Comment (RossComputerGuy) [reply pkgs/by-name/kr/krun/package.nix:20]: Used to semver, I'll change it.

File: pkgs/by-name/kr/krun/package.nix (hunk 4 above)
Comment (RossComputerGuy) [reply pkgs/by-name/kr/krun/package.nix:75]: 👍

File: pkgs/by-name/kr/krun/package.nix (hunk 5 above)
Comment (RossComputerGuy) [reply pkgs/by-name/kr/krun/package.nix:46]: Wanted to use `finalAttrs` but it doesn't seem `buildRustPackage` supports that?

Comment (Aleksanaa) [reply pkgs/by-name/kr/krun/package.nix:46]: Is `finalAttrs` really that important here? We don't use anything except `finalAttrs.src` which [does not work better than `rec`](https://github.com/NixOS/nixpkgs/issues/310373)

Comment (RossComputerGuy) [reply pkgs/by-name/kr/krun/package.nix:46]: Alright, I'll change it

File: pkgs/by-name/kr/krun/package.nix (hunk 3 above)
Comment (RossComputerGuy) [reply pkgs/by-name/kr/krun/package.nix]: Done

File: pkgs/by-name/kr/krun/package.nix (hunk 4 above)
Comment (RossComputerGuy) [reply pkgs/by-name/kr/krun/package.nix]: Fixed

File: pkgs/by-name/kr/krun/package.nix (hunk 5 above)
Comment (RossComputerGuy) [reply pkgs/by-name/kr/krun/package.nix]: Migrated to using `buildRustPackage`

**Comment by RossComputerGuy** [comment]
Found out a rebase fixed `libkrun`'s hash.

**Comment by RossComputerGuy** [comment]
Ok, now it actually works after I juggle the hashes around. Seems they all got shuffled.

**Review by Aleksanaa — APPROVED** [review]
Looks good

**Comment by Aleksanaa** [comment]
7 packages built:
krun krunvm libkrun libkrun-sev libkrun-sev.dev libkrun.dev libkrunfw

**Comment by Aleksanaa** [comment]
```
=> NIXOS_CURRENT_SYSTEM=$(readlink -e /run/current-system/) krun bash
No IPv6 nameserver available for NDP/DHCPv6
```

So I've been stuck at here for about 10 minutes. Will check later

**Comment by RossComputerGuy** [comment]
> So I've been stuck at here for about 10 minutes. Will check later

Maybe try spinning up a paast socket or a different command?

> ```
> --passt-socket=PATH  Instead of starting passt, connect to passt socket at PATH
> ```

**Comment by tomberek** [comment]
Result of `nixpkgs-review pr 321359` run on x86_64-linux [1](https://github.com/Mic92/nixpkgs-review)
<details>
  <summary>7 packages built:</summary>
  <ul>
    <li>krun</li>
    <li>krunvm</li>
    <li>libkrun</li>
    <li>libkrun-sev</li>
    <li>libkrun-sev.dev</li>
    <li>libkrun.dev</li>
    <li>libkrunfw</li>
  </ul>
</details>

**Review by tomberek — APPROVED** [review]

**Pull request merged**